torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,223 @@
+            // Child - None
+            // Append All Child Outputs to Parent Initial
+            initial_.insert(child_outputs.begin(), child_outputs.end());
+          } else if (child_sync_inserter.has_war_hazard_sync_) {
+            // Child - WAR race
+            // Parent first sync
+            // Inherit Child Initial / Clear Parent Final
+            initial_sync_ = true;
+            is_last_op_sync_ = true;
+            initial_.insert(
+                child_sync_inserter.initial().begin(),
+                child_sync_inserter.initial().end());
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Parent first sync
+            // Inherit Child Initial + Final
+            initial_sync_ = true;
+            initial_.insert(
+                child_sync_inserter.initial().begin(),
+                child_sync_inserter.initial().end());
+            final_.insert(
+                child_sync_inserter.final().begin(),
+                child_sync_inserter.final().end());
+          }
+        } else {
+          // Parent - 1+
+          if (!child_sync_inserter.initial_sync_) {
+            // Child - None
+            // Append All Child to Parent Last
+            final_.insert(child_inputs.begin(), child_inputs.end());
+          } else if (child_sync_inserter.has_war_hazard_sync_) {
+            // Child - WAR race
+            // Clear Parent Last / Discard Child Initial
+            is_last_op_sync_ = true;
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Inherit Child Final / Discard Child Initial
+            final_.insert(
+                child_sync_inserter.final().begin(),
+                child_sync_inserter.final().end());
+          }
+        }
+      } else {
+        handle(expr);
+      }
+    }
+
+    // This level of the nested for-loop may not exist in the kernel.
+    // However, subsequent levels can exist, so we handle the body of the
+    // for-loop first.
+    if (!fl->iter_domain()->isThread() && !fl->iter_domain()->isBroadcast()) {
+      // Determine if any smem TV is written to at beginning of the for-loop
+      // and whether that smem TV is read from at the end of the for-loop
+      size_t isize = intersection_size(initial_, final_);
+      // Insert new SyncThreads at end of for-loop to prevent WAR race condition
+      if (isize > 0 &&
+          fl->body().exprs().back()->getExprType().value() != ExprType::Sync &&
+          !is_last_op_sync_) {
+        // std::cout << "WAR race detected; Add Sync" << std::endl;
+        has_war_hazard_sync_ = true;
+        fl->body().push_back(new kir::Sync(true));
+      }
+    }
+  }
+
+  size_t intersection_size(
+      std::unordered_set<const TensorView*>& left,
+      std::unordered_set<const TensorView*>& right) {
+    size_t result = 0;
+    for (auto item : left) {
+      if (right.find(item) != right.end()) {
+        ++result;
```

Comment:
this is still more generic than it has to be: we just need a helper to tell is if the intersection is empty or now, which means we can return early on the first match.

Reply:
I was using the count earlier to check for correctness/debug.

Reply:
is it needed in the final version?